### DISCUSSION TITLE
Draft Proposal: Sum types and overloading for parametric polymorphism

### DISCUSSION THREAD

**Discussion by hez2010** [dotnet/csharplang#5085] [Language Ideas]
# Sum types and overloading for parametric polymorphism

## Summary
Support sum types and method overloading for parametric polymorphism.

## Motivation
Since abstract static methods in interface has been introduced, we can make use of generic math to generalize the implementation of number computation. But without sum types and method overloading support, we cannot specialize implementation for different kind of numbers (e.g. integer numbers and floating numbers). This also applies to scenarios other than generic math, and F# will also benefit from it.

## Detailed design
I'm splitting this feature into 3 parts.

### Method overloading by generic constraints
This will allow us to define method overloads with different generic constraints.

```csharp
public T Calc<T>(T value) where T : IBinaryIntegerNumber<T>
{
    // algorithm for integer numbers
}

public T Calc<T>(T value) where T : IBinaryFloatingNumber<T>
{
    // algorithm for floating numbers
}
```

This could be confusing: if a type `T` implements both `IBinaryIntegerNumber<T>` and `IBinaryFloatingNumber<T>`, which one should be called? Thus, we have the second part of proposal as following.

### Support 'or' on generic constraints
Today parametric polymorphism in C# only supports product types, for example:

```csharp
public T Calc<T>(T value) where T : IBinaryIntegerNumber<T>, IBinaryFloatingNumber<T>
{
    // algorithm for both integer and floating numbers
}
```

This part of the proposal will enable sum types for parametric polymorphism, and we can use it with type pattern matching together.

```csharp
public T Calc<T>(T value) where T : (IBinaryIntegerNumber<T> | IBinaryFloatingNumber<T>)
{
    switch (value)
    {
        case IBinaryIntegerNumber<T>: break;
        case IBinaryFloatingNumber<T>: break;
        // no default branch here since it's exhaustive
    }
}
```

Although this seems the same with method overloading by generic constraints in some ways, this approach eliminates the problem during method overload resolution.

### Support 'or' on type parameters for generic constraints
Currently in F#, if we define an `inline` method `add` like below:

```fsharp
let inline add a b = a + b
```

the method signature will be:

```fsharp
add :: x: ^a -> y: ^b -> ^c
    when (^a or ^b) : (static member ( + ) : ^a * ^b -> ^c)
```

This requires either `x` or `y` has a static member `+`, and the member `+` is a binary operator. But we can't express it neither in C# nor IL, so we cannot call that F# method from C# through ABI directly, and F# won't benefit from abstract static interface methods.

Therefore, I propose in C# we can also apply this `or` relation to type parameters:

```csharp
public T3 Add<T1, T2, T3>(T1 x, T2 y) where (T1 | T2) : IAdditionOperators<T1, T2, T3>
{
    return x + y;
}
```

### Drawbacks
This would definitely require a runtime type system change, but it worth and benefits all languages on .NET.

**Comment by YairHalberstadt** [comment]
Related: https://github.com/dotnet/csharplang/discussions/4628

**Reply by hez2010** [reply]
#4628 didn't take generics into account. Can this proposal be merged into #4628?
